pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ProducerCornerCaseTest.java

```diff
@@ -0,0 +1,356 @@
+ */
+package org.apache.pulsar.client.impl;
+
+import static org.awaitility.reflect.WhiteboxImpl.getInternalState;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertTrue;
+import io.netty.buffer.ByteBuf;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.pulsar.broker.BrokerTestUtil;
+import org.apache.pulsar.client.api.CompressionType;
+import org.apache.pulsar.client.api.Message;
+import org.apache.pulsar.client.api.MessageId;
+import org.apache.pulsar.client.api.Producer;
+import org.apache.pulsar.client.api.ProducerConsumerBase;
+import org.apache.pulsar.client.api.Schema;
+import org.apache.pulsar.client.api.interceptor.ProducerInterceptor;
+import org.awaitility.Awaitility;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+@Slf4j
+@Test(groups = "broker-api")
+public class ProducerCornerCaseTest extends ProducerConsumerBase {
+
+    @BeforeClass(alwaysRun = true)
+    @Override
+    protected void setup() throws Exception {
+        super.internalSetup();
+        super.producerBaseSetup();
+    }
+
+    @AfterClass(alwaysRun = true)
+    @Override
+    protected void cleanup() throws Exception {
+        super.internalCleanup();
+    }
+
+    @Override
+    protected void doInitConf() throws Exception {
+        super.doInitConf();
+        conf.setTopicLevelPoliciesEnabled(false);
+    }
+
+    @Test
+    public void testSendQueueIsFull() throws Exception {
+        final String topicName = BrokerTestUtil.newUniqueName("persistent://public/default/tp_");
+        admin.topics().createNonPartitionedTopic(topicName);
+        ProducerImpl<String> producer = (ProducerImpl<String>) pulsarClient.newProducer(Schema.STRING)
+                .blockIfQueueFull(false).maxPendingMessages(1)
+                .enableBatching(true).topic(topicName).create();
+        List<MsgPayloadTouchableMessageBuilder<String>> msgBuilderList = new ArrayList<>();
+        for (int i = 0; i < 100; i++) {
+            msgBuilderList.add(newMessage(producer));
+        }
+        int indexCalledSend = 0;
+        List<CompletableFuture> sendFutureList = new ArrayList<>();
+        for (MsgPayloadTouchableMessageBuilder<String> msgBuilder: msgBuilderList) {
+            try {
+                sendFutureList.add(msgBuilder.value("msg-1").sendAsync());
+                if (indexCalledSend != 99) {
+                    indexCalledSend++;
+                }
+            } catch (Exception ex) {
+                log.warn("", ex);
+                break;
+            }
+        }
+        try{
+            sendFutureList.get(sendFutureList.size() - 1).join();
+        } catch (Exception ex) {
+            log.warn("", ex);
+        }
```

Comment:
Which case is expected? Should it succeed or fail?

Reply:
Solved

Reply:
it's not solved. After I debugged this test, it should fail with `ProducerQueueIsFullError`, so I suggested the following change

```java
        try {
            sendFutureList.get(sendFutureList.size() - 1).join();
            fail();
        } catch (CompletionException e) {
            assertTrue(e.getCause() instanceof PulsarClientException.ProducerQueueIsFullError);
        }
```

Reply:
Accepted the suggestion and removed `fail()` to avoid flaky